Neos.Neos/Classes/Utility/NodeTypeWithFallbackProvider.php

```diff
@@ -0,0 +1,30 @@
+<?php
+
+namespace Neos\Neos\Utility;
+
+use Neos\ContentRepository\Core\NodeType\NodeType;
+use Neos\ContentRepository\Core\Projection\ContentGraph\Node;
+use Neos\ContentRepository\Core\SharedModel\Exception\NodeTypeNotFoundException;
+use Neos\ContentRepositoryRegistry\ContentRepositoryRegistry;
+use Neos\Flow\Annotations as Flow;
+use Neos\Neos\Domain\Service\NodeTypeNameFactory;
+
+/**
+ * Utility trait for retrieving node types for nodes with a built-in fallback mechanism
+ */
+trait NodeTypeWithFallbackProvider
+{
+    #[Flow\Inject]
+    protected ContentRepositoryRegistry $contentRepositoryRegistry;
```

Comment:
traits with flow inject annotation - hmmm something rings a bell @fcool said to me that `inject` here works more reliable and should we rename the property to something prefixed only for this trait like `$NodeTypeWithFallbackProvider_contentRepositoryRegistry`?

i did notice that currently you are removing the dependency from the classes where the trait is also injected

```diff
-    #[Flow\Inject]
-     protected ContentRepositoryRegistry $contentRepositoryRegistry;
```

i just dislike such intransitive dependencies ^^ or are my concerns are unfounded?

but good idea to have this trait <3

Reply:
you remember correctly - but this is only valid for the annotations.
Trouble with annotations is, that their namespace gets lost during proxy building. But attributes should be always fine.

Reply:
ah okay ;) and what about my concern of that this property pollutes now the class? I think it should only be visible to the trait (even though that is not possible - unless we obscurity the name ^^)

the trait is meant to be used in userland code... and i guess if another trait is using the same property name but without type and legacy annotation we have a problem?

Reply:
the registry injection has been extracted to a separate trait to clean this up